### PR TITLE
Video data transform

### PR DESCRIPTION
**Pull request by abeglova**
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [ ] Settings
  - [ ] New settings are documented and present in `app.json`
  - [ ] New settings have reasonable development defaults, if applicable

#### What are the relevant tickets?
Closes https://github.com/mitodl/open-discussions/issues/2304

#### What's this PR do?
This pr creates a transform function for normalizing youtube video data and adds a youtube pipeline

#### How should this be manually tested?
```
from course_catalog.etl import youtube
from toolz import compose

pipeline = compose(youtube.transform, youtube.extract)
g = pipeline()
next(g)
```
should step through the videos in the channels in YOUTUBE_CONFIG_FILE_LOCATION displaying normalized data for one video at a time

```
from course_catalog.etl.pipelines import *
youtube_etl()
```
should pull all the videos in the channels in YOUTUBE_CONFIG_FILE_LOCATION and save them in the database

#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

#### What GIF best describes this PR or how it makes you feel?
(Optional)

### CODE REVIEW

File: course_catalog/etl/pipelines_test.py [hunk 1]
```diff
@@ -109,3 +109,19 @@ def test_oll_etl(mocker):
     mock_load_courses.assert_called_once_with(mock_transform.return_value)
 
     assert result == mock_load_courses.return_value
+
+
+def test_youtube_etl(mocker):
+    """Verify that youtube etl pipeline executes correctly"""
+    mock_extract = mocker.patch("course_catalog.etl.youtube.extract")
+    mock_transform = mocker.patch("course_catalog.etl.youtube.transform")
+    mock_load_courses = mocker.patch("course_catalog.etl.loaders.load_videos")
```
Comment (rhysyngsun) [inline course_catalog/etl/pipelines_test.py]:
```suggestion
    mock_load_videos = mocker.patch("course_catalog.etl.loaders.load_videos")
```

File: course_catalog/etl/pipelines_test.py [hunk 2]
```diff
@@ -109,3 +109,19 @@ def test_oll_etl(mocker):
...
+
+    with reload_mocked_pipeline(mock_extract, mock_transform, mock_load_courses):
```
Comment (rhysyngsun) [inline course_catalog/etl/pipelines_test.py]:
```suggestion
    with reload_mocked_pipeline(mock_extract, mock_transform, mock_load_videos):
```

File: course_catalog/etl/pipelines_test.py [hunk 3]
```diff
@@ -109,3 +109,19 @@ def test_oll_etl(mocker):
...
+        result = pipelines.youtube_etl()
+
+    mock_extract.assert_called_once_with()
+    mock_transform.assert_called_once_with(mock_extract.return_value)
+    mock_load_courses.assert_called_once_with(mock_transform.return_value)
```
Comment (rhysyngsun) [inline course_catalog/etl/pipelines_test.py]:
```suggestion
    mock_load_videos.assert_called_once_with(mock_transform.return_value)
```

File: course_catalog/etl/pipelines_test.py [hunk 4]
```diff
@@ -109,3 +109,19 @@ def test_oll_etl(mocker):
...
+
+    assert result == mock_load_courses.return_value
```
Comment (rhysyngsun) [inline course_catalog/etl/pipelines_test.py]:
```suggestion
    assert result == mock_load_videos.return_value
```

File: course_catalog/etl/youtube.py [hunk 5]
```diff
@@ -101,3 +102,41 @@ def extract():
...
+    Args:
+        offered_by (string)
+        raw_video_data (dict)
+
+    Returns:
+        dict: normalized video data
+    """
+    return {
+        "video_id": raw_video_data["id"],
+        "platform": offered_by,
```
Comment (rhysyngsun) [inline course_catalog/etl/youtube.py]: This should be `PlatformType.youtube.value`. For `offered_by`, it needs to be a list of dicts, look at xPro as an example.

File: course_catalog/etl/youtube.py [hunk 6]
```diff
@@ -101,3 +102,41 @@ def extract():
...
+        "full_description": raw_video_data["snippet"]["description"],
```
Comment (rhysyngsun) [inline course_catalog/etl/youtube.py]: I think we also need a title here

File: course_catalog/etl/youtube.py [hunk 7]
```diff
@@ -101,3 +102,41 @@ def extract():
...
+        "image_src": raw_video_data["snippet"]["thumbnails"]["high"]["url"],
+        "last_modified": raw_video_data["snippet"]["publishedAt"],
+        "published": True,
+        "url": "https://www.youtube.com/watch?v=" + raw_video_data["id"],
```
Comment (rhysyngsun) [inline course_catalog/etl/youtube.py]: This should use a format string instead of a string concat

File: course_catalog/etl/youtube.py [hunk 8]
```diff
@@ -101,3 +102,41 @@ def extract():
...
+    """
+    Transforms raw video data into normalized data structure
+
+    Args:
+        iterable that yields raw video data in the the form of (offered_by, data) tuples
+
+    Returns:
+        generator that yields normalized video data
+    """
+    for video_data in raw_videos_data:
```
Comment (rhysyngsun) [inline course_catalog/etl/youtube.py]: To make this a little more clear what's being passed into `transform_single_video`, Python lets you destructure a tuple in a for loop like `for offered_by, raw_video_data in raw_videos_data:`, then you can pass each param into the call

File: course_catalog/etl/youtube.py [hunk 9]
```diff
@@ -101,3 +102,41 @@ def extract():
...
+    }
+
+
+@log_exceptions("Error transforming youtube data", exc_return_value=[])
+def transform(raw_videos_data):
+    """
+    Transforms raw video data into normalized data structure
+
+    Args:
+        iterable that yields raw video data in the the form of (offered_by, data) tuples
```
Comment (rhysyngsun) [inline course_catalog/etl/youtube.py]: The arg syntax format should have the variable name and type like this:
```suggestion
        raw_videos_data (iterable of tuples) tuple has the structure (offered_by, data)
```

File: course_catalog/etl/pipelines_test.py (hunk 1 above)
Comment (abeglova) [reply course_catalog/etl/pipelines_test.py]: oops fixed!

**Pull request merged**